lib/metrician/middleware/request_timing.rb

```diff
@@ -17,96 +17,129 @@ def call(env)
         process_start_time = Time.now.to_f
         response_size = 0
 
-        queue_start_time = self.class.extract_request_start_time(env)
-        gauge("queue_time", process_start_time - queue_start_time) if queue_start_time
+        if Middleware.queue_time?
+          queue_start_time = self.class.extract_request_start_time(env)
+          gauge(:queue_time, process_start_time - queue_start_time) if queue_start_time
+        end
 
-        if @request_end_time
-          gauge("idle", process_start_time - @request_end_time)
-          @request_end_time = nil
+        if Middleware.idle?
+          if @request_end_time
+            gauge(:idle, process_start_time - @request_end_time)
+            @request_end_time = nil
+          end
         end
 
         begin
           status, headers, body = @app.call(env)
-          response_size = self.class.get_response_size(headers: headers, body: body)
           [status, headers, body]
         ensure
-          current_route = self.class.extract_route(
-            controller: env["action_controller.instance"],
-            path: env["REQUEST_PATH"]
-          )
-          request_time = env["REQUEST_TOTAL_TIME"].to_f
-          env["REQUEST_TOTAL_TIME"] = nil
-          gauge("request", request_time, current_route)
-          apdex(request_time)
-          increment("error", current_route) if status >= 500
-
-          # Note that 30xs don't have content-length, so cached
-          # items will report other metrics but not this one
-          if response_size && !response_size.to_s.strip.empty?
-            gauge("response_size", response_size.to_i, current_route)
+          if need_route?
+            current_route = self.class.extract_route(
+              controller: env[ENV_CONTROLLER_PATH],
+              path: env[ENV_REQUEST_PATH]
+            )
+            if Middleware.request?
+              request_time = env[ENV_REQUEST_TOTAL_TIME].to_f
+              env[ENV_REQUEST_TOTAL_TIME] = nil
+              gauge(:request, request_time, current_route)
+            end
+            if Middleware.apdex?
+              apdex(request_time)
```

Comment:
It seems like this might explode if `Middleware.request?` returns false because `request_time` wouldn't be defined. If so, maybe we need to set `request_time` if `Middleware.request? || Middleware.apdex?`, and probably add a test.